omegaconf/grammar_visitor.py

```diff
@@ -136,16 +136,26 @@ def visitInterpolation(
     def visitInterpolationNode(
         self, ctx: OmegaConfGrammarParser.InterpolationNodeContext
     ) -> Optional["Node"]:
-        # INTER_OPEN DOT* (configKey (DOT configKey)*)? INTER_CLOSE
+        #   INTER_OPEN DOT*                                            // dots for relative interpolations
+        #   (
+        #     configKey                                                // first config key
+        #     (DOT configKey | BRACKET_OPEN configKey BRACKET_CLOSE)*  // .foo, [foo], .foo[bar], [foo].bar[baz]
+        #   )?
+        #   INTER_CLOSE;
         assert ctx.getChildCount() >= 2
 
         inter_key_tokens = []  # parsed elements of the dot path
         for child in ctx.getChildren():
             if isinstance(child, TerminalNode):
-                if child.symbol.type == OmegaConfGrammarLexer.DOT:
-                    inter_key_tokens.append(".")  # preserve dots
+                if child.symbol.type in (
+                    OmegaConfGrammarLexer.DOT,
+                    OmegaConfGrammarLexer.BRACKET_OPEN,
+                ):
+                    # Always use dot notation for simplicity.
+                    inter_key_tokens.append(".")
```

Comment:
I am assuming that this is why you did not have to change `_resolve_key_and_root()`.
One thing I am suspecting that this will cause some unexpected error messages where the input contains brackets and the error message is in pure dot notation.

Reply:
> I am assuming that this is why you did not have to change `_resolve_key_and_root()`.

No, I didn't change `_resolve_key_and_root()` because as far as I can tell it only relies on the `.` notation for the "relative" part of the interpolation (i.e., the initial dots). And I didn't touch that part.

> One thing I am suspecting that this will cause some unexpected error messages where the input contains brackets and the error message is in pure dot notation.

Indeed! Do you believe it matters? If yes I can preserve the bracket notation instead.

Reply:
Can you show an a short session that results in such an error?
Will help get a feeling of how bad it looks.

Reply:
```python
cfg = OmegaConf.create({
    "x": "${foo[a][b]}",
})
cfg.x
```

```
---------------------------------------------------------------------------
InterpolationKeyError                     Traceback (most recent call last)
(...)
InterpolationKeyError: Interpolation key 'foo.a.b' not found
    full_key: x
    object_type=dict
```

Reply:
I changed it in 8131646 to show the correct path (it was important to keep the brackets to support https://github.com/omry/omegaconf/pull/647#discussion_r602629863)